### ISSUE TITLE
[Announcement] Status of the project

### ISSUE THREAD

**Issue by jspenguin2017** [AdBlockProtector2/AdBlockProtector2#5]
Hi everyone, I see some people are staring and watching this repository, I'll just give everyone a heads-up on what are my plans and what are current problems.
First, ABR2 has 2 major dependencies: uBR and libmicro. Currently, libmicro is no where close to being functional, and as I will not be using it, I have little incentive to write it. Unless I can learn something from it, there will hardly be any progress in the near future. Therefore, I will be rewriting libmicro in TypeScript, with Gulp for building. It's the first TypeScript project for me, I pretty much have no idea what I'm doing right now, hopefully I'll know better soon. 
Second, ABR2 is an experiment, just like AAK-Cont and unlike uBR, I will not be using it myself, and I'll be honest with you, I probably won't care too much if it is slow or it break. Sure I want it to be shiny and fast, but that's easier said than done, pretty much everything else in my life will have higher priority than fixing ABR2. If there is a bug that affect only one website for one browser that's not Chrome, be ready to wait months before it gets resolved.

**Comment by sir-harlekin** [comment]
Hey there,

thanks for your information. I guess since AAK-Cont is dicontinued, you are our only hope.

Please keep up your great work.


Kind regards

**Comment by jspenguin2017** [comment]
**Update**

Initial implementation of libmicro is ready! `$replace=<trusted_replace_handler>` is not yet implemented, I plan on adding `Micro.setReplacer` which takes in a hash table of functions to be replace handlers. It will only work on Quantum though. 

Currently, libmicro can only handle `$inject=<trusted_scriptlet>` and `$redirect=<trusted_resource>` along with general resource blocking. 

Example test filters: 
```
||test.example.com^$document,important
||test2.example.com^$libmicro,redirect=hello-world.html
||example.com^$libmicro,inject=hello-world.js
```

**Comment by jspenguin2017** [comment]
OK, that's not going really well, the current implementation is really inefficient, although 2k filters shouldn't choke it. 
There are 3 choices: 
* Just get it up and running, and deal with performance later
* Fork uBlock Origin
* Rewrite it again in WebAssembly

**Comment by sir-harlekin** [comment]
Well, how about get it up and running first? There will be support coming in then.

**Comment by jspenguin2017** [comment]
I don't think there will be any significant coding support from the community, that didn't happen last time, probably won't happen this time neither.

**Comment by jspenguin2017** [comment]
I think WebAssembly could be worth a shot. It'll be very useful in so many ways if I can pull it off. 
But to get things up fast and smooth, forking uBO is probably the safest bet.

**Comment by jspenguin2017** [comment]
I'll test out my ideas with wasm, it's definitely worth a shot. 
https://github.com/AdBlockProtector2/libpico

**Comment by jspenguin2017** [comment]
Forget what I said, I'm forking uBO.

**Comment by Wetitpig** [comment]
So for Quantum what are we going to use now? AAK is discountinued already and somehow is not very effective.

Btw, will the new adblock protector support adguard?

**Comment by jspenguin2017** [comment]
No, Adguard doesn't need it. Report problems to their forum instead.

**Comment by Wetitpig** [comment]
Adguard is definitely easily detected by anti adblocks. I will force install and see how it goes.

**Comment by jspenguin2017** [comment]
You should report any detection to Adguard's issues tracker, they don't know it, they can't fix it.

**Comment by jspenguin2017** [comment]
It takes an absolute eternity to get an uBO fork though Web Store. Well, my plan doesn't work, like always... 
Looks like we might have to distribute it on our own. Can be really difficult to get it though Windows security.

**Comment by jspenguin2017** [comment]
@Wetitpig Also, AFAIK, Adguard got into some legal trouble and had to disable some anti-adblock defusers by default, make sure to activate the annoyance filter.

**Comment by jspenguin2017** [comment]
I find myself more interested in working on an adblocker than for an adblocker, I might cancel ABR2. If someone wants to take over, let me know.

**Comment by elypter** [comment]
i would love to see a version of ublock that is more open to new features. but i also imagine that it is going to be a lot of work. if you think you can do it then go for it. however still try to get in contact with gorhill. he isn't particularly easy to work with and he had bad experiences in the past but you could still try to work in parallel so nobody is putting stones in each others way and occasionally post pull requests in both directions if that is desired.

**Comment by jspenguin2017** [comment]
If you put yourself in his shoe, you'll understand why he isn't open to new features, if it breaks, it breaks for well over 20 million people, if my code break, maybe a few hundreds people are affected. 

Some of the suggested features are REALLY hard to implement, it may seem easy, but after 1h you realize that you have 2 bugs but you can only fix 1 without rewriting thousands of lines of code. This is exactly what happened to https://github.com/NanoAdblocker/NanoCore/issues/12

But we should never settle, without a fork there won't be progress. I am open to new features, but lots of them are kind of hard to implement. I put in a syntax highlighter, which should be a pretty big plus, but making a linter is another story. 

uBO is over-optimized, and can't be extended easily.

**Comment by elypter** [comment]
i think the best way is to prioritise features that are easy to add and do not require too many changes. this makes merging with the gorhills code easier as well. i like that performance is a high priority in uBo but i rather have a slow but working feature than a fast non existing one. i suggest to go through the uBo issue tracker and search for all closed and denied or closed feature request issues and then bookmark everything that you think is doable. if you get stuck with one feature put it on halt and move over to the next thing. get back to it later. sometimes ideas and new perspectives come with time.

**Comment by jspenguin2017** [comment]
That's what I'm doing. I'm rapidly putting simple yet powerful features into Nano. 1.0.0.6 is probably the best adblocker for filter maintainers. Syntax highlighting so you know you typed it right, and more debugging hooks to pin point the problem. 
I plan on adding "comment section" and "uncomment section" shortcut keys, and auto-complete. For now, Nano is aiming at developers.

**Comment by elypter** [comment]
i love the extension of script snippets. i hope that NanoBlock can one day replace the now discontinued cleanlinks addon or what decentral eyes is doing and if it can do more replacement tasks instead of classic deleting.

i am currently writing a userscript that gets rid of all those annoying scrolling headers and footers. https://github.com/elypter/BlockHead
turns out that to make it efficient i would probably have to write a whole new adblocker from scratch. it would be great if tasks like this could work together with an adblocker to prevent duplicate code and code execution.

**Comment by jspenguin2017** [comment]
I can add more script snippets. I'm not very sure what clean link does, but after a quick look, I think it's implementable in script snippets. Although I would need a better issue report in order to properly implement it, open issues in Nano's repo if you have something in mind. 
For scrolling header and footer you can try `:style(` or `addEventListener-defuser.js`, give it a try and see if you can achieve what you want. If you need more script snippets, let me know and I'll see what I can do.

**Comment by FrostKnight** [comment]
Would it be at all possible to support firefox esr even if its limited support till june 2018? I greatly support what your doing, I just wondered if you would be willing do to that.
There are still people using firefox esr.  people from distros such as debian and ubuntu use firefox esr. As well as organizations too, Ublock Origin is what I use.

**Comment by jspenguin2017** [comment]
ABR2 is probably going to be canceled. 
Nano Adblocker might support legacy Firefox, but not now. However, the support will be immediately dropped when it's dropped in uBO. 
Nano Defender will not support legacy browsers, it might support Firefox Quantum 58 or 59+

**Comment by FrostKnight** [comment]
On 12/12/2017 08:45 AM, jspenguin2017 wrote:
> ABR2 is probably going to be canceled. 
> Nano Adblocker might support legacy Firefox, but not now. However, the support will be immediately dropped when it's dropped in uBO. 
> Nano Defender will not support legacy browsers, it might support Firefox Quantum 58 or 59+
>
>
Well, until June 2018 would be what would be required to support esr...
no longer, no less.

Due to 2018 being the last esr version without web extensions.

At that moment, it becomes pointless as you would say to support legacy
addons.

ps, can I get the same functionality following this?
https://jspenguin2017.github.io/uBlockProtector/

and how long will it last if it does?

**Comment by jspenguin2017** [comment]
I don't know how to write legacy extensions. And as it is reaching EOL soon, I'm not going to learn it.

**Comment by FrostKnight** [comment]
On 12/12/2017 07:25 PM, jspenguin2017 wrote:
> I don't know how to write legacy extensions. And as it is reaching EOL soon, I'm not going to learn it. 
>
That's fine.  It's not like its going to last much longer like you just
said. Is there a place to donate to you if I wanted to?

**Comment by jspenguin2017** [comment]
I don't accept donations. 
Accepting donations means accepting responsibility, I want to be free to do what I want and not be bound to maintaining a project.

**Comment by FrostKnight** [comment]
On 12/12/2017 08:16 PM, jspenguin2017 wrote:
> I don't accept donations. 
> Accepting donations means accepting responsibility, I want to be free to do what I want and not be bound to maintaining a project. 
>
>
That makes sense.  Well, thank you anyways for making this project.
